File: superai/harvester.py
"""Self-learning harvester - the loop that makes Super AI improve itself.

A background thread that, forever:
  1. scrapes GitHub trending + raw README/code files and learns them
  2. resolves the "curiosity queue" (topics users asked about that the
     brain didn't know) via Wikipedia + GitHub search scraping
  3. retrains the LlamaLite neural model on the grown corpus
  4. logs every step to the evolution feed

Pure stdlib scraping (urllib + regex). No AI API anywhere.
"""

import json
import random
import re
import threading
import time
import urllib.parse
import urllib.request
from typing import Dict, List, Optional

UA = {"User-Agent": "SuperAI-SelfLearner/1.0 (+self-training research bot)"}

SEED_SOURCES = [
    ("https://en.wikipedia.org/wiki/Artificial_intelligence", "text"),
    ("https://en.wikipedia.org/wiki/Machine_learning", "text"),
    ("https://en.wikipedia.org/wiki/Large_language_model", "text"),
    ("https://raw.githubusercontent.com/torvalds/linux/master/README", "text"),
    ("https://raw.githubusercontent.com/python/cpython/main/README.rst", "text"),
]

GITHUB_TRENDING = "https://github.com/trending"


def fetch(url: str, timeout: int = 15) -> str:
    req = urllib.request.Request(url, headers=UA)
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return resp.read().decode("utf-8", errors="ignore")


def strip_html(html: str) -> str:
    html = re.sub(r"<script[^>]*>.*?</script>", " ", html, flags=re.S | re.I)
    html = re.sub(r"<style[^>]*>.*?</style>", " ", html, flags=re.S | re.I)
    text = re.sub(r"<[^>]+>", " ", html)
    text = re.sub(r"&[a-z]+;", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def extract_title(html: str) -> str:
    m = re.search(r"<title>(.*?)</title>", html, re.I | re.S)
    return re.sub(r"\s+", " ", m.group(1)).strip()[:150] if m else ""


def extract_code_blocks(html: str) -> List[str]:
    blocks = re.findall(r"<(?:code|pre)[^>]*>(.*?)</(?:code|pre)>", html, re.S | re.I)
    out = []
    for b in blocks:
        b = strip_html(b)
        if 30 < len(b) < 800:
            out.append(b)
    return out[:10]


class Harvester:
    """Autonomous learning loop. Runs as a daemon thread."""

    def __init__(self, brain, state, interval: int = 600):
        self.brain = brain
        self.state = state
        self.interval = interval
        self.enabled = True
        self.last_run: Optional[float] = None
        self.last_report: Dict = {}
        self._thread: Optional[threading.Thread] = None

    # ------------------------------------------------------------------

    def learn_url(self, url: str) -> Dict:
        """Learn from one URL right now (also used by the /api/learn route)."""
        if not url.startswith(("http://", "https://")):
            return {"ok": False, "error": "URL must start with http:// or https://"}
        if self.state.has_source(url):
            return {"ok": True, "learned": 0, "note": "already learned this source"}
        try:
            raw = fetch(url)
        except Exception as e:
            return {"ok": False, "error": f"fetch failed: {e}"}

        is_raw_file = "raw.githubusercontent.com" in url or not raw.lstrip().startswith("<")
        if is_raw_file:
            title = url.rsplit("/", 1)[-1]
            kind = "code" if re.search(r"\.(py|js|ts|go|rs|java|c|cpp|rb|sh)$", url) else "text"
            n = self.brain.learn_text(url, title, raw[:15000], kind=kind)
            self.brain.evolve("learned-url", f"{url} -> {n} sentences ({kind})")
            return {"ok": True, "learned": n, "title": title}

        title = extract_title(raw)
        text = strip_html(raw)[:15000]
        n = self.brain.learn_text(url, title, text, kind="text")
        for i, code in enumerate(extract_code_blocks(raw)):
            n += self.brain.learn_text(f"{url}#code{i}", title, code, kind="code")
        self.brain.evolve("learned-url", f"{url} -> {n} sentences")
        return {"ok": True, "learned": n, "title": title}

    def harvest_github_trending(self) -> int:
        """Scrape GitHub trending, then learn READMEs of top repos."""
        learned = 0
        try:
            html = fetch(GITHUB_TRENDING)
        except Exception:
            return 0
        repos = re.findall(r'href="/([\w.-]+/[\w.-]+)"\s', html)
        seen = []
        for r in repos:
            if r.count("/") == 1 and r not in seen and not r.startswith(("login", "features", "trending", "topics", "sponsors", "site", "about", "contact", "pricing")):
                seen.append(r)
        for repo in seen[:3]:
            for branch in ("main", "master"):
                url = f"https://raw.githubusercontent.com/{repo}/{branch}/README.md"
                if self.state.has_source(url):
                    break
                try:
                    body = fetch(url)
                except Exception:
                    continue
                n = self.brain.learn_text(url, f"GitHub: {repo}", body[:12000], kind="code")
                if n:
                    learned += n
                    self.brain.evolve("github-harvest", f"learned {repo} README ({n} sentences)")
                break
        return learned

    def resolve_curiosity(self) -> int:
        """Self-learn topics users asked about but the brain didn't know."""
        queue = self.state.get_json("curiosity_queue", [])
        if not queue:
            return 0
        topic = queue.pop(0)
        self.state.put_json("curiosity_queue", queue)
        learned = 0
        # Wikipedia article scrape
        slug = urllib.parse.quote(topic.replace(" ", "_").title())
        url = f"https://en.wikipedia.org/wiki/{slug}"
        try:
            raw = fetch(url)
            text = strip_html(raw)[:12000]
            if len(text) > 500:
                learned += self.brain.learn_text(url, f"Wikipedia: {topic}", text)
        except Exception:
            pass
        if learned:
            self.brain.evolve("curiosity-resolved",
                              f"self-learned '{topic}' ({learned} sentences)")
        return learned

    # ------------------------------------------------------------------

    def seed_if_empty(self):
        if self.state.doc_count() > 0:
            return
        self.brain.evolve("boot", "first boot - seeding initial knowledge")
        for url, kind in SEED_SOURCES:
            try:
                raw = fetch(url)
                title = extract_title(raw) or url.rsplit("/", 1)[-1]
                text = strip_html(raw)[:12000] if raw.lstrip().startswith("<") else raw[:12000]
                n = self.brain.learn_text(url, title, text, kind=kind)
                self.brain.evolve("seed-learn", f"{url} -> {n} sentences")
            except Exception:
                continue
        # first neural training on the seed corpus
        self.brain.train_neural(steps=150)

    def cycle(self) -> Dict:
        """One full self-improvement cycle."""
        report = {"at": time.time(), "github": 0, "curiosity": 0, "neural": None}
        report["github"] = self.harvest_github_trending()
        report["curiosity"] = self.resolve_curiosity()
        if report["github"] or report["curiosity"] or random.random() < 0.5:
            report["neural"] = self.brain.train_neural(steps=120)
        self.last_run = time.time()
        self.last_report = report
        return report

    def _loop(self):
        # seed on boot (in background so the server starts instantly)
        try:
            self.seed_if_empty()
        except Exception:
            pass
        while True:
            time.sleep(self.interval)
            if not self.enabled:
                continue
            try:
                self.cycle()
            except Exception:
                pass

    def start(self):
        self._thread = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()

    def status(self) -> Dict:
        return {
            "enabled": self.enabled,
            "interval_sec": self.interval,
            "last_run": self.last_run,
            "last_report": self.last_report,
        }

File: assets/js/harvester.js
// harvester.js — the 24×7 self-learning loop, running in the browser.
//
// GitHub Pages is static, so learning happens client-side via fetch() against
// CORS-friendly sources only:
//   * raw.githubusercontent.com  (READMEs / raw code — sends ACAO: *)
//   * en.wikipedia.org REST API  (article summaries — CORS enabled)
//   * api.github.com             (repo search — CORS enabled)
// Anything a user asked that the brain didn't know goes on a curiosity queue
// and gets researched automatically.

import { LANGUAGES } from "./knowledge.js";

// Seed docs across many languages (all CORS-friendly raw files).
const SEED = [
  ["https://raw.githubusercontent.com/python/cpython/main/README.rst", "code"],
  ["https://raw.githubusercontent.com/nodejs/node/main/README.md", "code"],
  ["https://raw.githubusercontent.com/microsoft/TypeScript/main/README.md", "code"],
  ["https://raw.githubusercontent.com/golang/go/master/README.md", "code"],
  ["https://raw.githubusercontent.com/rust-lang/rust/master/README.md", "code"],
  ["https://raw.githubusercontent.com/torvalds/linux/master/README", "text"],
];

const LANG_TOPICS = {
  Python: "Python (programming language)", JavaScript: "JavaScript", TypeScript: "TypeScript",
  Java: "Java (programming language)", "C++": "C++", Go: "Go (programming language)",
  Rust: "Rust (programming language)", Ruby: "Ruby (programming language)",
  Swift: "Swift (programming language)", Kotlin: "Kotlin (programming language)",
  SQL: "SQL", Haskell: "Haskell", Scala: "Scala (programming language)",
};

function stripHtml(html) {
  html = html.replace(/<script[\s\S]*?<\/script>/gi, " ").replace(/<style[\s\S]*?<\/style>/gi, " ");
  return html.replace(/<[^>]+>/g, " ").replace(/&[a-z]+;/g, " ").replace(/\s+/g, " ").trim();
}

async function fetchText(url, timeoutMs = 15000) {
  const ctrl = new AbortController();
  const t = setTimeout(() => ctrl.abort(), timeoutMs);
  try {
    const res = await fetch(url, { signal: ctrl.signal, headers: { Accept: "text/plain,*/*" } });
    if (!res.ok) throw new Error("HTTP " + res.status);
    return await res.text();
  } finally {
    clearTimeout(t);
  }
}

export class Harvester {
  constructor(brain, store, intervalSec = 300) {
    this.brain = brain;
    this.store = store;
    this.intervalSec = intervalSec;
    this.enabled = true;
    this.lastRun = null;
    this.lastReport = {};
    this._timer = null;
    this.onUpdate = null; // callback to refresh UI
  }

  // Learn one URL right now (also used by "Teach from URL").
  async learnUrl(url) {
    if (!/^https?:\/\//.test(url)) return { ok: false, error: "URL must start with http:// or https://" };
    if (await this.store.hasSource(url)) return { ok: true, learned: 0, note: "already learned this source" };
    let raw;
    try { raw = await fetchText(url); }
    catch (e) { return { ok: false, error: `fetch failed (${e.message}). Note: many sites block cross-origin requests; raw.githubusercontent.com and Wikipedia work.` }; }

    const isRawFile = url.includes("raw.githubusercontent.com") || !raw.trimStart().startsWith("<");
    let title, kind, text;
    if (isRawFile) {
      title = url.split("/").pop();
      kind = /\.(py|js|ts|go|rs|java|c|cpp|rb|sh|kt|swift|php|sql)$/.test(url) ? "code" : "text";
      text = raw.slice(0, 15000);
    } else {
      const m = raw.match(/<title>([\s\S]*?)<\/title>/i);
      title = m ? m[1].replace(/\s+/g, " ").trim().slice(0, 150) : url;
      kind = "text";
      text = stripHtml(raw).slice(0, 15000);
    }
    const n = await this.brain.learnText(url, title, text, kind);
    await this.brain.evolve("learned-url", `${url} -> ${n} sentences (${kind})`);
    return { ok: true, learned: n, title };
  }

  async harvestGitHub() {
    // search a random popular language, learn a top repo's README
    const lang = LANGUAGES[Math.floor(Math.random() * LANGUAGES.length)];
    let learned = 0;
    try {
      const api = `https://api.github.com/search/repositories?q=language:${encodeURIComponent(lang)}&sort=stars&order=desc&per_page=5`;
      const data = JSON.parse(await fetchText(api));
      for (const repo of (data.items || []).slice(0, 3)) {
        for (const branch of ["main", "master"]) {
          const url = `https://raw.githubusercontent.com/${repo.full_name}/${branch}/README.md`;
          if (await this.store.hasSource(url)) break;
          try {
            const body = await fetchText(url);
            const n = await this.brain.learnText(url, `GitHub: ${repo.full_name}`, body.slice(0, 12000), "code");
            if (n) { learned += n; await this.brain.evolve("github-harvest", `${lang}: learned ${repo.full_name} (${n} sentences)`); }
            break;
          } catch { /* try next branch */ }
        }
        if (learned) break;
      }
    } catch { /* GitHub API rate-limited or offline */ }
    return learned;
  }

  async resolveCuriosity() {
    const queue = await this.store.getJSON("curiosity_queue", []);
    if (!queue.length) return 0;
    const topic = queue.shift();
    await this.store.setJSON("curiosity_queue", queue);
    let learned = 0;
    const title = (LANG_TOPICS[topic] || topic).replace(/\b\w/g, (c) => c.toUpperCase());
    const url = `https://en.wikipedia.org/api/rest_v1/page/summary/${encodeURIComponent(title.replace(/ /g, "_"))}`;
    try {
      const data = JSON.parse(await fetchText(url));
      if (data.extract && data.extract.length > 120) {
        learned = await this.brain.learnText(
          data.content_urls?.desktop?.page || url, `Wikipedia: ${data.title}`, data.extract, "text");
        await this.brain.evolve("curiosity-resolved", `self-learned "${topic}" (${learned} sentences)`);
      }
    } catch { /* not found / offline */ }
    return learned;
  }

  async seedIfEmpty() {
    if ((await this.store.docCount()) > 40) return; // KB already seeded ~45
    for (const [url, kind] of SEED) {
      try {
        const body = await fetchText(url);
        const title = url.split("/").pop();
        await this.brain.learnText(url, title, body.slice(0, 12000), kind);
        await this.brain.evolve("seed-learn", `${title} -> learned`);
      } catch { /* offline is fine — KB still works */ }
    }
  }

  async cycle() {
    const report = { at: Date.now(), github: 0, curiosity: 0, neural: null };
    try { report.github = await this.harvestGitHub(); } catch {}
    try { report.curiosity = await this.resolveCuriosity(); } catch {}
    if (report.github || report.curiosity || Math.random() < 0.6) {
      try { report.neural = await this.brain.trainNeural(120); } catch {}
    }
    this.lastRun = Date.now();
    this.lastReport = report;
    if (this.onUpdate) this.onUpdate();
    return report;
  }

  start() {
    // background seed + first training, then loop forever while the tab is open
    (async () => {
      try { await this.seedIfEmpty(); } catch {}
      if (this.brain.llama.stepsTrained === 0) {
        try { await this.brain.trainNeural(150); if (this.onUpdate) this.onUpdate(); } catch {}
      }
    })();
    const tick = async () => {
      if (this.enabled) { try { await this.cycle(); } catch {} }
      this._timer = setTimeout(tick, this.intervalSec * 1000);
    };
    this._timer = setTimeout(tick, this.intervalSec * 1000);
  }

  status() {
    return { enabled: this.enabled, interval_sec: this.intervalSec, last_run: this.lastRun, last_report: this.lastReport };
  }
}
